python/lsst/ts/m2com/biquadratic_filter.py

```diff
@@ -0,0 +1,182 @@
+# This file is part of ts_m2com.
+#
+# Developed for the Vera Rubin Observatory Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["SingleBiquadraticFilter", "BiquadraticFilter"]
+
+import numpy as np
+import numpy.typing
+
+
+class SingleBiquadraticFilter:
+    """Single biquadratic filter.
+
+    Notes
+    -----
+    The equation is:
+
+           1 + b11 z^-1 + b21 z^-2    Y(z)
+    H(z) = ----------------------- = -----
+           1 + a11 z^-1 + a21 z^-2.   X(z)
+
+    This can be considered as the following process:
+
+    g[n-2] = b21 * x[n] - a21 * y[n]
+    g[n-1] = g[n-2] + b11 * x[n] - a11 * y[n]
+
+    Therefore, we will have:
+
+    y[n] = x[n] + g[n-1]
+
+    Parameters
+    ----------
+    a11 : `float`
+        The coefficient a11 in the equation.
+    a21 : `float`
+        The coefficient a21 in the equation.
+    b11 : `float`
+        The coefficient b11 in the equation.
+    b21 : `float`
+        The coefficient b21 in the equation.
+    num_element : `int`
+        Number of the elements.
+    """
+
+    def __init__(
+        self, a11: float, a21: float, b11: float, b21: float, num_element: int
+    ) -> None:
+        self._a11 = a11
+        self._a21 = a21
+        self._b11 = b11
+        self._b21 = b21
+
+        self._g_m1 = np.zeros(num_element)
+        self._g_m2 = np.zeros(num_element)
+
+    def reset(self) -> None:
+        """Reset the filter."""
+
+        self._g_m1 = np.zeros(len(self._g_m1))
+        self._g_m2 = np.zeros(len(self._g_m2))
+
+    def filter(
+        self, value: numpy.typing.NDArray[np.float64]
+    ) -> numpy.typing.NDArray[np.float64]:
+        """Filter the input value.
+
+        Parameters
+        ----------
+        value : `numpy.ndarray`
+            Input value.
+
+        Returns
+        -------
+        value_filter : `numpy.ndarray`
+            Filtered value.
+        """
+
+        # Filtered value
+        value_filter = value + self._g_m1
```

Comment:
There's an implicit requirement here that the input `value` array must have the same dimension of `self._g_m1`. 

Maybe document this in the docstring? I don't think adding a check here would be necessary, as numpy would raise an exception. But documenting this might be useful. You may want to say that the class needs to be initialized with the same number of elements etc...

Reply:
Improved in the commit: [e30ff64](https://github.com/lsst-ts/ts_m2com/pull/49/commits/e30ff642d7ddd6c36fd869fa856d6163921ee184). Thanks!